dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/PredictionContextGenerator.java

```diff
@@ -0,0 +1,157 @@
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.predictor;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.hisp.dhis.category.CategoryOptionCombo;
+import org.hisp.dhis.common.DimensionalItemObject;
+import org.hisp.dhis.common.FoundDimensionItemValue;
+import org.hisp.dhis.common.MapMap;
+import org.hisp.dhis.common.MapMapMap;
+import org.hisp.dhis.period.Period;
+
+/**
+ * Generator of prediction contexts.
+ *
+ * @author Jim Grace
+ */
+public class PredictionContextGenerator
+{
+    private PredictionContextGenerator() // static methods only
+    {
+    }
+
+    /**
+     * Generates prediction contexts. Each prediction context contains all the
+     * values required to evaluate a predictor to (possibly) generate one
+     * prediction.
+     * <p>
+     * All the data used to generate contexts has the same organisation unit.
+     *
+     * @param outputPeriods output periods (predict within each period)
+     * @param values input prediction values (all with the same orgUnit)
+     * @param defaultCategoryOptionCombo system default cat option combo
+     * @return contexts for prediction evaluation
+     */
+    public static List<PredictionContext> getContexts( List<Period> outputPeriods,
+        List<FoundDimensionItemValue> values, CategoryOptionCombo defaultCategoryOptionCombo )
+    {
+        List<PredictionContext> contexts = new ArrayList<>();
+
+        MapMapMap<CategoryOptionCombo, Period, DimensionalItemObject, Object> aocMap = getAocMap(
```

Comment:
In general I agree, and here in particular I tried to think of ways to improve on using the `MapMapMap`. However, every solution I came up with made the code less concise and IMHO less readable. If all we had to do was to reference the objects later I agree completely, and this could be done with a key class, or even just a method that returns a `String` key that represents all the dimensions. I've done exactly this elsewhere.

The reason this `MapMapMap` is so useful is the way it unwinds in the `getContexts` method: (1) we need to iterate over each Attribute Option Combination to generate a number of contexts for each AOC which are only known by looking at the found data. (2) For each AOC, we need a `MapMap<Period, DimensionalItemObject, Object> periodValueMap` ultimately to send to `ExpressionService` to fetch the `valueMap` for each sampled period (unless we want to extend the refactoring into `ExpressionService` as well, which I'm open to but just now I want to get this bug fixed). (3) For each output period, we need a `Map<DimensionalItemObject, Object> valueMap` again for `ExpressionService` to give the values for the non-sampled data. Again, maybe this could be refactored in the `ExpressionService`, although there may be some good reasons not to, such as only prediction expressions deal with sampled periods while all other types of expressions just use this simple `valueMap`.

Each solution I could think of for using a simple `Map` involved more objects to use for the looping to create the contexts, and significantly more code logic to do the job. I'm open to changing the code, but I'm not yet satisfied that eliminating `MapMapMap` will result in easier code to understand and maintain. The conceptual difficulty of understanding `MapMapMap` is less than the difficulty of understanding an alternative I can think of so far, IMHO.

Keep in mind also that we are looking at only data that is found, not a complete set of all the dimensions. Not every possible AOC may be found in the data. Not every sampled `Period` may be found in the data. Not every `DimensionalItemObject` may be in the data. The `MapMapMap` provides a way of assembling the data in a structured way and then unwinding it very succinctly into the collections needed.

Let me know if I've missed your point, which is entirely possible. :)

Reply:
Didn't quite get this fully but I think I understand the important part that you make use of the maps with 2 and 1 key component in a way that would be hard with special key classes. Fair enough.